### PR TITLE
REPL Ctrl-D termination, unclosed string handling and other refactors

### PR DESCRIPTION
**Pull request by alexmozaidze**
After I've read your [blog](https://www.brandons.me/blog/favorite-rust-function), I decided to check on `parser.rs` and found out how simple it actually is, which amazed me. I like the blogs, keep it up.

I did following things in this PR:
- Made REPL close on Ctrl-D (at least on Linux)
- Handled unclosed strings properly so it doesn't panic
- Allowed single semicolon comments
- Bumped Rust edition (2018 -> 2021)
- Did some general refactoring

### CODE REVIEW

**Comment by alexmozaidze** [comment]
Should we update the edition to 2021? It seems that it compiles just fine, and tests complete without any issues.

**Comment by alexmozaidze** [comment]
Rebasing grants me so much power, I feel like a time traveler.

**Comment by brundonsmith** [comment]
Some thoughts:
- I'm not a fan of `clippy::clone_on_ref_ptr` and the related changes; let's revert those
- I'm not seeing where it checks for Ctrl+D; does `.lines()` do that automatically?
- I think we want to return `None` instead of `Some(&code[index..])` on line 81 in `parser.rs`. At least in the REPL this bubbles up as an `Unrecognized expression` instead of automatically closing the string, which I think is better.

Otherwise I think these changes look good

**Comment by brundonsmith** [comment]
Ah, one other thing- there are some `|c: char|`s in here that don't need the type annotation

**Comment by alexmozaidze** [comment]
Yeah, `.lines()` returns an iterator, and it's `None` when Ctrl-D is hit. I was quite surprised it worked so well.

**Comment by alexmozaidze** [comment]
I did the changes and simplified `read_atom()` by ommiting unclosed strings check.

File: src/parser.rs [hunk 1]
```diff
@@ -372,9 +374,14 @@ fn read_atom(token: &str) -> Value {
...
     }
 
-    if token.chars().next().map_or(false, |c| c == '"')
-        && token.chars().nth_back(0).map_or(false, |c| c == '"')
-    {
+    if token.chars().next().map_or(false, |c| c == '"') {
+        // Empty string edge-case
+        if token.chars().nth(1).map_or(false, |c| c == '"') {
+            return Value::String(String::from(""));
+        }
```
Comment (brundonsmith) [inline src/parser.rs:381]: I don't think this edge-case is needed. I added a test for parsing an empty string, and then just deleted this whole clause and it still worked fine

Comment (alexmozaidze) [reply src/parser.rs:381]: Hmm, that's wierd. When I entered an empty string (`""`) in the REPL it came out as (`"""`)

Comment (alexmozaidze) [reply src/parser.rs:381]: Oooooh, if you do `x[a..a]` aka it's a range from a value to the same value, it will give out an empty string as long as we're in a valid range. That's... interesting.

**Comment by brundonsmith** [comment]
Merged and published to crates.io
